File: examples/strategies/rsi_threshold/python/runner.py
"""Market-data and order orchestration for the RSI threshold example."""

from __future__ import annotations

import logging
from collections.abc import Sequence
from datetime import datetime, timedelta, timezone
from decimal import Decimal

from google.protobuf.timestamp_pb2 import Timestamp
from google.type.decimal_pb2 import Decimal as ProtoDecimal
from google.type.interval_pb2 import Interval
from trade_api import TradeAPIClient
from trade_api.accounts import GetAccountRequest
from trade_api.auth_messages import TokenDetailsRequest
from trade_api.market_data import Bar, BarsRequest, SubscribeBarsRequest
from trade_api.orders import Order, OrderStatus, OrderType, Side, TimeInForce

from config import TIMEFRAMES, Config
from strategy import HISTORY_CLOSES, MINIMUM_CLOSES, Signal, evaluate

logger = logging.getLogger(__name__)

_DISPLAY = Decimal("0.000001")


def _bar_key(bar: Bar) -> tuple[int, int]:
    return bar.timestamp.seconds, bar.timestamp.nanos


def ordered_bars(bars: Sequence[Bar]) -> list[Bar]:
    """Sort bars and keep the newest update for each timestamp."""

    by_timestamp = {_bar_key(bar): bar for bar in bars}
    return [by_timestamp[timestamp] for timestamp in sorted(by_timestamp)]


def _decimal(value: ProtoDecimal) -> Decimal:
    return Decimal(value.value or "0")


def _format(value: Decimal) -> str:
    """Round a value for display only; the rule uses the full value."""

    return str(value.quantize(_DISPLAY))


def _history(client: TradeAPIClient, config: Config) -> tuple[list[Decimal], Bar]:
    timeframe, lookback_days = TIMEFRAMES[config.timeframe]
    end = datetime.now(timezone.utc)
    start_timestamp, end_timestamp = Timestamp(), Timestamp()
    start_timestamp.FromDatetime(end - timedelta(days=lookback_days))
    end_timestamp.FromDatetime(end)

    response = client.market_data.Bars(
        BarsRequest(
            symbol=config.symbol,
            timeframe=timeframe,
            interval=Interval(start_time=start_timestamp, end_time=end_timestamp),
        )
    )
    bars = ordered_bars(response.bars)
    required = MINIMUM_CLOSES + 1
    if len(bars) < required:
        raise RuntimeError(
            f"at least {required} historical bars are required; received {len(bars)}"
        )

    # Keep the newest bar pending because it may still be changing.
    return [_decimal(bar.close) for bar in bars[:-1]], bars[-1]


def _position(client: TradeAPIClient, account_id: str, symbol: str) -> Decimal:
    account = client.accounts.GetAccount(GetAccountRequest(account_id=account_id))
    match = next((item for item in account.positions if item.symbol == symbol), None)
    return _decimal(match.quantity) if match else Decimal(0)


def resolve_account_id(client: TradeAPIClient) -> str:
    token = client.get_token() or ""
    details = client.auth.TokenDetails(TokenDetailsRequest(token=token))
    if len(details.account_ids) != 1:
        raise RuntimeError(
            f"expected exactly one available account; received {len(details.account_ids)}"
        )
    return str(details.account_ids[0])


def place_order(
    client: TradeAPIClient,
    config: Config,
    account_id: str | None,
    signal: Signal,
    bar: Bar,
) -> bool:
    """Act on a signal and report whether to adopt the position it implies.

    A dry run only logs, but still adopts, so the simulated position tracks the
    rule. An entry blocked by a guard is not adopted, which keeps the strategy
    flat and free to act on the next signal. An exit blocked because the account
    holds nothing is adopted, because the guard has just confirmed that the
    strategy is flat; an order that never filled cannot strand it.
    """

    if signal is None:
        return False
    if not config.execute:
        logger.warning("DRY RUN: %s %s units of %s", signal, config.quantity, config.symbol)
        return True
    if not account_id:
        raise RuntimeError("account ID is required in execute mode")

    current = _position(client, account_id, config.symbol)
    if signal == "entry" and current != 0:
        logger.warning("Skipping entry: current position is %s, expected zero", current)
        return False
    if signal == "exit" and current <= 0:
        logger.warning("Skipping exit: there is no long position")
        return True

    side = Side.SIDE_BUY if signal == "entry" else Side.SIDE_SELL
    quantity = config.quantity if signal == "entry" else min(current, config.quantity)
    suffix = str(bar.timestamp.seconds)[-10:]
    state = client.orders.PlaceOrder(
        Order(
            account_id=account_id,
            symbol=config.symbol,
            quantity=ProtoDecimal(value=format(quantity, "f")),
            side=side,
            type=OrderType.ORDER_TYPE_MARKET,
            time_in_force=TimeInForce.TIME_IN_FORCE_DAY,
            client_order_id=f"rsi14-{'b' if signal == 'entry' else 's'}-{suffix}",
            comment="RSI threshold",
        )
    )
    logger.warning("Submitted order %s: %s", state.order_id, OrderStatus.Name(state.status))
    return True


def run(client: TradeAPIClient, config: Config) -> None:
    account_id = resolve_account_id(client) if config.execute else None
    closes, pending = _history(client, config)
    result = evaluate(closes)
    logger.info("History ready: close=%s rsi=%s", closes[-1], _format(result.rsi))

    if config.check:
        print(
            f"History check passed: close={closes[-1]} rsi={_format(result.rsi)} "
            f"average_gain={_format(result.average_gain)} "
            f"average_loss={_format(result.average_loss)} signal={result.signal or 'none'}"
        )
        return

    # The strategy starts flat and never adopts a position it did not open.
    in_position = False
    timeframe = TIMEFRAMES[config.timeframe][0]
    request = SubscribeBarsRequest(symbol=config.symbol, timeframe=timeframe)
    for response in client.market_data.SubscribeBars(request):
        for bar in ordered_bars(response.bars):
            if _bar_key(bar) < _bar_key(pending):
                continue
            if _bar_key(bar) == _bar_key(pending):
                pending = bar
                continue

            closes = [*closes, _decimal(pending.close)][-HISTORY_CLOSES:]
            result = evaluate(closes, in_position)
            logger.info(
                "Closed bar: close=%s rsi=%s average_gain=%s average_loss=%s signal=%s",
                closes[-1],
                _format(result.rsi),
                _format(result.average_gain),
                _format(result.average_loss),
                result.signal or "none",
            )
            if place_order(client, config, account_id, result.signal, pending):
                in_position = result.signal == "entry"
            pending = bar

File: examples/strategies/rsi_threshold/node/runner.ts
import type { TradeApi } from "@limeint/trade-api";
import type { Bar } from "@limeint/trade-api/market-data";
import { OrderType, orderStatusToJSON, Side, TimeInForce } from "@limeint/trade-api/orders";
import type { Config, LogLevel } from "./config.js";
import { TIMEFRAMES } from "./config.js";
import type { Signal } from "./strategy.js";
import { evaluate, HISTORY_CLOSES, MINIMUM_CLOSES } from "./strategy.js";

export type StrategyApi = Readonly<{
  auth: Pick<TradeApi["auth"], "tokenDetails">;
  getToken: TradeApi["getToken"];
  marketData: Pick<TradeApi["marketData"], "bars" | "subscribeBars">;
  accounts: Pick<TradeApi["accounts"], "getAccount">;
  orders: Pick<TradeApi["orders"], "placeOrder">;
}>;

const LOG_PRIORITY: Record<LogLevel, number> = {
  DEBUG: 10,
  INFO: 20,
  WARNING: 30,
  ERROR: 40,
};

export const log = (config: Config, level: LogLevel, message: string): void => {
  if (LOG_PRIORITY[level] < LOG_PRIORITY[config.logLevel]) return;
  if (level === "ERROR") console.error(message);
  else if (level === "WARNING") console.warn(message);
  else console.info(message);
};

const barKey = (bar: Pick<Bar, "timestamp">): number => {
  if (!bar.timestamp) throw new Error("bar timestamp is missing");
  return bar.timestamp.getTime();
};

export const orderedBars = (bars: readonly Bar[]): Bar[] => {
  const byTimestamp = new Map<number, Bar>();
  for (const bar of bars) byTimestamp.set(barKey(bar), bar);
  return [...byTimestamp.entries()].sort(([left], [right]) => left - right).map(([, bar]) => bar);
};

const decimal = (value: { value: string } | undefined): number => {
  const parsed = Number(value?.value ?? "0");
  if (!Number.isFinite(parsed)) throw new Error(`invalid decimal: ${value?.value}`);
  return parsed;
};

/** Round a value for display only; the rule uses the full value. */
const format = (value: number): string => value.toFixed(6);

const history = async (
  api: StrategyApi,
  config: Config,
  now = new Date(),
): Promise<{ closes: number[]; pending: Bar }> => {
  const [timeframe, lookbackDays] = TIMEFRAMES[config.timeframe];
  const response = await api.marketData.bars({
    symbol: config.symbol,
    timeframe,
    interval: {
      startTime: new Date(now.getTime() - lookbackDays * 86_400_000),
      endTime: now,
    },
  });
  const bars = orderedBars(response.bars);
  const required = MINIMUM_CLOSES + 1;
  if (bars.length < required) {
    throw new Error(`at least ${required} historical bars are required; received ${bars.length}`);
  }
  const pending = bars.at(-1);
  if (!pending) throw new Error("historical response did not contain a pending bar");

  return {
    closes: bars.slice(0, -1).map((bar) => decimal(bar.close)),
    pending,
  };
};

const position = async (api: StrategyApi, accountId: string, symbol: string): Promise<number> => {
  const account = await api.accounts.getAccount({ accountId });
  return decimal(account.positions.find((item) => item.symbol === symbol)?.quantity);
};

export const resolveAccountId = async (api: StrategyApi): Promise<string> => {
  const details = await api.auth.tokenDetails({ token: api.getToken() });
  const [accountId] = details.accountIds;
  if (details.accountIds.length !== 1 || !accountId) {
    throw new Error(
      `Expected exactly one available account; received ${details.accountIds.length}`,
    );
  }
  return accountId;
};

/**
 * Act on a signal and report whether to adopt the position it implies.
 *
 * A dry run only logs, but still adopts, so the simulated position tracks the
 * rule. An entry blocked by a guard is not adopted, which keeps the strategy
 * flat and free to act on the next signal. An exit blocked because the account
 * holds nothing is adopted, because the guard has just confirmed that the
 * strategy is flat; an order that never filled cannot strand it.
 */
export const placeOrder = async (
  api: StrategyApi,
  config: Config,
  accountId: string | undefined,
  signal: Signal,
  bar: Bar,
): Promise<boolean> => {
  if (!signal) return false;
  if (!config.execute) {
    log(config, "WARNING", `DRY RUN: ${signal} ${config.quantity} units of ${config.symbol}`);
    return true;
  }

  if (!accountId) throw new Error("account ID is required in execute mode");

  const current = await position(api, accountId, config.symbol);
  if (signal === "entry" && current !== 0) {
    log(config, "WARNING", `Skipping entry: current position is ${current}, expected zero`);
    return false;
  }
  if (signal === "exit" && current <= 0) {
    log(config, "WARNING", "Skipping exit: there is no long position");
    return true;
  }

  const side = signal === "entry" ? Side.SIDE_BUY : Side.SIDE_SELL;
  const quantity = signal === "entry" ? config.quantity : Math.min(current, config.quantity);
  const seconds = Math.floor(barKey(bar) / 1_000)
    .toString()
    .slice(-10)
    .padStart(10, "0");
  const state = await api.orders.placeOrder(
    {
      accountId,
      symbol: config.symbol,
      quantity: { value: String(quantity) },
      side,
      type: OrderType.ORDER_TYPE_MARKET,
      timeInForce: TimeInForce.TIME_IN_FORCE_DAY,
      clientOrderId: `rsi14-${signal === "entry" ? "b" : "s"}-${seconds}`,
      comment: "RSI threshold",
    },
    { retry: false },
  );
  log(config, "WARNING", `Submitted order ${state.orderId}: ${orderStatusToJSON(state.status)}`);
  return true;
};

type MarketState = {
  closes: number[];
  pending: Bar;
  inPosition: boolean;
};

const processBar = async (
  api: StrategyApi,
  config: Config,
  state: MarketState,
  bar: Bar,
  accountId: string | undefined,
): Promise<MarketState> => {
  if (barKey(bar) < barKey(state.pending)) return state;
  if (barKey(bar) === barKey(state.pending)) return { ...state, pending: bar };

  const closes = [...state.closes, decimal(state.pending.close)].slice(-HISTORY_CLOSES);
  const result = evaluate(closes, state.inPosition);
  log(
    config,
    "INFO",
    `Closed bar: close=${closes.at(-1)} rsi=${format(result.rsi)} ` +
      `average_gain=${format(result.averageGain)} average_loss=${format(result.averageLoss)} ` +
      `signal=${result.signal ?? "none"}`,
  );
  const changed = await placeOrder(api, config, accountId, result.signal, state.pending);
  return {
    closes,
    pending: bar,
    inPosition: changed ? result.signal === "entry" : state.inPosition,
  };
};

export const run = async (api: StrategyApi, config: Config): Promise<void> => {
  const accountId = config.execute ? await resolveAccountId(api) : undefined;
  // The strategy starts flat and never adopts a position it did not open.
  let state: MarketState = { ...(await history(api, config)), inPosition: false };
  const result = evaluate(state.closes);
  log(config, "INFO", `History ready: close=${state.closes.at(-1)} rsi=${format(result.rsi)}`);

  if (config.check) {
    console.log(
      `History check passed: close=${state.closes.at(-1)} rsi=${format(result.rsi)} ` +
        `average_gain=${format(result.averageGain)} average_loss=${format(result.averageLoss)} ` +
        `signal=${result.signal ?? "none"}`,
    );
    return;
  }

  const [timeframe] = TIMEFRAMES[config.timeframe];
  for await (const response of api.marketData.subscribeBars({
    symbol: config.symbol,
    timeframe,
  })) {
    for (const bar of orderedBars(response.bars)) {
      state = await processBar(api, config, state, bar, accountId);
    }
  }
};
